// filepond/esm/svelte/svelte/src/internal/client/reactivity/props.js
import { PROPS_IS_UPDATED as E, PROPS_IS_BINDABLE as R, PROPS_IS_IMMUTABLE as L, PROPS_IS_LAZY_INITIAL as T } from "../../../constants.js";
import { get_descriptor as P, is_function as c } from "../../shared/utils.js";
import { set as B } from "./sources.js";
import { derived as O, derived_safe_equal as Y } from "./deriveds.js";
import { get as _, is_destroying_effect as j, active_effect as K, untrack as M } from "../runtime.js";
import { props_invalid_value as N } from "../errors.js";
import { DESTROYED as U, STATE_SYMBOL as I, LEGACY_PROPS as x } from "../constants.js";
import { proxy as $ } from "../proxy.js";
import { capture_store_binding as q } from "./store.js";
const z = {
  get(e, r) {
    if (!e.exclude.includes(r))
      return e.props[r];
  },
  set(e, r) {
    return !1;
  },
  getOwnPropertyDescriptor(e, r) {
    if (!e.exclude.includes(r) && r in e.props)
      return {
        enumerable: !0,
        configurable: !0,
        value: e.props[r]
      };
  },
  has(e, r) {
    return e.exclude.includes(r) ? !1 : r in e.props;
  },
  ownKeys(e) {
    return Reflect.ownKeys(e.props).filter((r) => !e.exclude.includes(r));
  }
};
// @__NO_SIDE_EFFECTS__
function k(e, r, t) {
  return new Proxy(
    { props: e, exclude: r },
    z
  );
}
const C = {
  get(e, r) {
    let t = e.props.length;
    for (; t--; ) {
      let n = e.props[t];
      if (c(n) && (n = n()), typeof n == "object" && n !== null && r in n) return n[r];
    }
  },
  set(e, r, t) {
    let n = e.props.length;
    for (; n--; ) {
      let i = e.props[n];
      c(i) && (i = i());
      const u = P(i, r);
      if (u && u.set)
        return u.set(t), !0;
    }
    return !1;
  },
  getOwnPropertyDescriptor(e, r) {
    let t = e.props.length;
    for (; t--; ) {
      let n = e.props[t];
      if (c(n) && (n = n()), typeof n == "object" && n !== null && r in n) {
        const i = P(n, r);
        return i && !i.configurable && (i.configurable = !0), i;
      }
    }
  },
  has(e, r) {
    if (r === I || r === x) return !1;
    for (let t of e.props)
      if (c(t) && (t = t()), t != null && r in t) return !0;
    return !1;
  },
  ownKeys(e) {
    const r = [];
    for (let t of e.props)
      if (c(t) && (t = t()), !!t) {
        for (const n in t)
          r.includes(n) || r.push(n);
        for (const n of Object.getOwnPropertySymbols(t))
          r.includes(n) || r.push(n);
      }
    return r;
  }
};
function ee(...e) {
  return new Proxy({ props: e }, C);
}
function re(e, r, t, n) {
  var i = !0, u = (t & R) !== 0, h = (t & T) !== 0, a = (
    /** @type {V} */
    n
  ), v = !0, g = (
    /** @type {Derived<V> | undefined} */
    void 0
  ), S = () => h && i ? (g ??= O(
    /** @type {() => V} */
    n
  ), _(g)) : (v && (v = !1, a = h ? M(
    /** @type {() => V} */
    n
  ) : (
    /** @type {V} */
    n
  )), a);
  let o;
  if (u) {
    var y = I in e || x in e;
    o = P(e, r)?.set ?? (y && r in e ? (f) => e[r] = f : void 0);
  }
  var s, b = !1;
  u ? [s, b] = q(() => (
    /** @type {V} */
    e[r]
  )) : s = /** @type {V} */
  e[r], s === void 0 && n !== void 0 && (s = S(), o && (N(), o(s)));
  var l;
  if (l = () => {
    var f = (
      /** @type {V} */
      e[r]
    );
    return f === void 0 ? S() : (v = !0, f);
  }, (t & E) === 0)
    return l;
  if (o) {
    var A = e._$legacy;
    return (
      /** @type {() => V} */
      (function(f, d) {
        return arguments.length > 0 ? ((!d || A || b) && o(d ? l() : f), f) : l();
      })
    );
  }
  var m = !1, p = ((t & L) !== 0 ? O : Y)(() => (m = !1, l()));
  u && _(p);
  var D = (
    /** @type {Effect} */
    K
  );
  return (
    /** @type {() => V} */
    (function(f, d) {
      if (arguments.length > 0) {
        const w = d ? _(p) : u ? $(f) : f;
        return B(p, w), m = !0, a !== void 0 && (a = w), f;
      }
      return j && m || (D.f & U) !== 0 ? p.v : _(p);
    })
  );
}
export {
  re as prop,
  k as rest_props,
  ee as spread_props
};
