// filepond/esm/elements/FilePondEntryList/components/MediaImage/components/BitmapRenderer.svelte.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { onMount as H, onDestroy as E } from "../../../../../svelte/svelte/src/index-client.js";
import { user_effect as w } from "../../../../../svelte/svelte/src/internal/client/reactivity/effects.js";
import { push as G, pop as J } from "../../../../../svelte/svelte/src/internal/client/context.js";
import { get as e } from "../../../../../svelte/svelte/src/internal/client/runtime.js";
import { state as l, set as o } from "../../../../../svelte/svelte/src/internal/client/reactivity/sources.js";
import { from_html as K, append as N } from "../../../../../svelte/svelte/src/internal/client/dom/template.js";
import { user_derived as v } from "../../../../../svelte/svelte/src/internal/client/reactivity/deriveds.js";
import { bind_this as V } from "../../../../../svelte/svelte/src/internal/client/dom/elements/bindings/this.js";
import { prop as d } from "../../../../../svelte/svelte/src/internal/client/reactivity/props.js";
import { isFirefox as X } from "../../../../../utils/test.js";
import { didAbort as Y } from "../../../../../utils/abort.js";
import { createObjectURL as Z } from "../../../../../utils/objectURL.js";
import { getImageSize as $ } from "../../../../../utils/media.js";
import { getAppContext as ee } from "../../../contexts/appContext.js";
import { getBitmapCacheItem as te, setBitmapCacheItem as z } from "./BitmapRendererCache.js";
import { thread as ie, createThreadWorker as re } from "../../../../../utils/thread.js";
import { transformImage as ae } from "../../../../../workers/transformImage.js";
var oe = K('<canvas width="0" height="0"></canvas>');
function xe(B, t) {
  G(t, !0);
  let k = d(t, "taskIgnoreSoftFailure", 3, !1), g = d(t, "maximumPixels", 3, 1024 * 1024), D = d(t, "resizeQuality", 3, "medium"), O = d(t, "workersURL", 3, void 0), T = d(t, "oninit", 3, void 0), R = d(t, "onload", 3, void 0), W = d(t, "onrender", 3, void 0), y = d(t, "onerror", 3, void 0);
  const j = () => Z(t.file), x = () => t.file.type !== "image/svg+xml", { size: h, canvas: c } = te(t.file) ?? {};
  let r = h?.width, a = h?.height, S = l(!!h), b = l(!!h), I = l(!!c), u = l(!1);
  w(() => {
    e(u) && W()?.({ didRestore: !!c });
  });
  const { pushTask: C, abortTask: L } = ee();
  let f = l(void 0), i = l(void 0);
  const q = v(() => !!e(i)), P = v(() => e(q) && !e(S)), Q = v(() => e(q) && e(b) && !e(I));
  async function A(n, { signal: s }) {
    o(I, !0);
    try {
      const m = await ie(
        re(O(), ae),
        [
          t.file,
          null,
          {
            resizeWidth: r,
            resizeHeight: a,
            resizeQuality: D(),
            imageOrientation: "from-image"
          }
        ],
        { signal: s }
      );
      e(i).width = r, e(i).height = a, e(i).getContext("bitmaprenderer")?.transferFromImageBitmap(m), z(t.file, { size: { width: r, height: a }, canvas: e(i) }), o(u, !0);
    } catch (m) {
      if (Y(s, m))
        throw m;
      y()?.(m);
    }
  }
  async function p() {
    o(I, !0), o(f, new Image()), e(f).src = j(), await e(f).decode(), e(i).width = r, e(i).height = a, e(i).getContext("2d")?.drawImage(e(f), 0, 0, r, a), z(t.file, { size: { width: r, height: a }, canvas: e(i) }), o(u, !0);
  }
  const F = X() || !x() ? p : A;
  async function M() {
    o(S, !0);
    let n;
    try {
      n = await $(t.file);
    } catch (_) {
      throw y()?.(_), _;
    }
    let s = 1;
    const m = n.width * n.height;
    g() && (m > g() || !x()) && (s = Math.sqrt(g()) / Math.sqrt(m)), r = Math.floor(n.width * s), a = Math.floor(n.height * s), R()?.({ width: r, height: a }), e(i).width = r, e(i).height = a, z(t.file, { size: { width: r, height: a }, canvas: null }), o(b, !0);
  }
  w(() => {
    e(P) && C(t.taskId, M, { parallel: 1, ignoreSoftFailure: k() });
  }), w(() => {
    e(Q) && C(t.taskId, F, { parallel: 1, ignoreSoftFailure: k() });
  }), H(() => {
    T()?.(), h && (R()?.({ width: r, height: a }), c && (e(i).replaceWith(c), o(u, !0)));
  }), E(() => {
    L(t.taskId, M), L(t.taskId, F), e(f) && URL.revokeObjectURL(e(f).src);
  });
  var U = oe();
  V(U, (n) => o(i, n), () => e(i)), N(B, U), J();
}
export {
  xe as default
};
