// filepond/esm/extensions/chunked-upload-store.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createStoreExtension as H } from "./common/createStoreExtension.js";
import { isString as g, isFileEntry as I, isFile as K, isNumber as z } from "../utils/test.js";
import { didAbort as x } from "../utils/abort.js";
import { naturalFileSizeToBytes as L } from "../utils/file.js";
import { passthrough as k, noop as $ } from "../utils/placeholder.js";
import { sleep as J } from "../utils/sleep.js";
import { createProgressEvent as N, xhr as S, getResponseHeaders as j } from "../utils/xhr.js";
import { warn as G } from "../common/console.js";
const oe = H({
  name: "ChunkedUploadStore",
  props: {
    url: "",
    chunkSize: 1 / 0,
    retryDelays: [500, 1e3, 3e3],
    parallelChunks: 2,
    resume: !1,
    resolveRequest: {},
    resolveResponse: {}
  },
  factory: ({ props: v, didSetProps: U }, { updateEntry: P }) => {
    let q = 1 / 0;
    U(({ chunkSize: e }) => {
      if (q = g(e) ? L(e) : e || 1 / 0, q <= 1024) {
        G("Chunk size has to be more than 1 kilobyte");
        return;
      }
    });
    function E(e, o) {
      const { valueKey: s } = v;
      P(e, {
        state: {
          [s]: o
        }
      });
    }
    function b(e) {
      const o = [], s = Math.floor(e.size / q);
      for (let a = 0; a <= s; a++) {
        const u = a * q || 0, l = e.slice(
          u,
          Math.min(u + q, e.size),
          "application/offset+octet-stream"
        );
        o.push({
          index: a,
          offset: u,
          size: l.size,
          data: l
        });
      }
      return o;
    }
    async function F(e, o, s) {
      const {
        url: a,
        resolveRequest: { create: u = k },
        resolveResponse: { create: l = ({ value: i }) => i }
      } = v, { signal: n } = o ?? {}, f = await u({
        url: a,
        options: {
          method: "POST",
          headers: {
            uploadLength: e.size
          }
        },
        entry: s
      }), r = await S(f.url, {
        ...f.options,
        signal: n
      }), t = l({
        response: r,
        value: r.response,
        entry: s
      });
      if (!g(t) || !t.length)
        throw new Error("No server id returned");
      return t;
    }
    async function O(e, o, s) {
      const {
        url: a,
        resolveRequest: { status: u = k },
        resolveResponse: { status: l = ({ value: p }) => p }
      } = v, { signal: n } = o ?? {}, f = await u({
        url: a,
        options: {
          method: "HEAD",
          queryString: {
            id: e
          }
        },
        entry: s
      }), r = await S(f.url, {
        ...f.options,
        signal: n
      }), { uploadOffset: t } = j(r), i = l({
        response: r,
        value: {
          id: e,
          offset: g(t) && t.length ? parseInt(t, 10) : Number.NaN
        },
        entry: s
      }), m = z(i.offset) && !Number.isNaN(i.offset), d = Array.isArray(i.chunks);
      if (!m && !d)
        throw new Error("No upload status returned");
      return i;
    }
    async function T(e, o, { uploadName: s, uploadLength: a }, u, l) {
      const {
        url: n,
        retryDelays: f,
        resolveRequest: { chunk: r = k },
        resolveResponse: { chunk: t = ({ value: d }) => d }
      } = v, { onprogress: i, signal: m } = u ?? {};
      for (const d of [...f, void 0])
        try {
          const p = await r({
            url: n,
            options: {
              method: "PATCH",
              headers: {
                uploadOffset: e.offset,
                uploadName: s,
                uploadLength: a
              },
              data: e.data,
              queryString: {
                contentType: "application/offset+octet-stream",
                id: o
              }
            },
            entry: l,
            chunk: e
          }), y = await S(p.url, {
            ...p.options,
            onprogress: i,
            signal: m
          });
          return t({
            response: y,
            value: {
              index: e.index,
              offset: e.offset,
              size: e.size
            },
            entry: l,
            id: o,
            chunk: e
          });
        } catch (p) {
          if (x(m, p))
            throw p;
          if (z(d))
            await J(d);
          else
            throw p;
        }
      throw new Error("Chunk upload failed");
    }
    async function A(e, o, { uploadName: s, uploadLength: a }, u, l) {
      const { parallelChunks: n } = v, { onprogress: f = $, signal: r } = u ?? {}, t = [], i = parseInt(a, 10), m = e.reduce((c, h) => c + h.size, 0);
      let d = Math.max(0, i - m);
      const p = /* @__PURE__ */ new Map(), y = () => {
        const c = Array.from(p.values()).reduce(
          (R, C) => R + C,
          0
        ), h = Math.min(d + c, i);
        f(N(!0, h, i));
      }, B = (c, h) => {
        h.lengthComputable && (p.set(c.index, Math.min(h.loaded, c.size)), y());
      };
      if (r?.aborted)
        throw r.reason;
      let w = [];
      const D = (c, h) => {
        const R = h.then((C) => {
          p.delete(c.index), d += c.size, t.push(C), y();
        }).finally(() => {
          w = w.filter((C) => C !== R);
        });
        w.push(R);
      };
      y();
      for (const c of e) {
        if (r?.aborted)
          throw r.reason;
        const h = T(
          c,
          o,
          {
            uploadName: s,
            uploadLength: a
          },
          {
            onprogress: (R) => B(c, R),
            signal: r
          },
          l
        );
        D(c, h), !(w.length < n) && await Promise.race(w);
      }
      return await Promise.all(w), t.sort((c, h) => c.index - h.index);
    }
    async function M(e, o, s, a) {
      const {
        url: u,
        resolveRequest: { complete: l },
        resolveResponse: { complete: n = ({ value: m }) => m }
      } = v, { signal: f } = s ?? {};
      if (!l)
        return e;
      const r = await l({
        url: u,
        options: {
          method: "POST",
          data: JSON.stringify({ chunks: o }),
          queryString: {
            id: e
          }
        },
        entry: a,
        id: e,
        chunks: o
      }), t = await S(r.url, {
        ...r.options,
        signal: f
      }), i = n({
        response: t,
        value: g(t.response) && t.response.length ? t.response : e,
        entry: a,
        id: e,
        chunks: o
      });
      if (!g(i) || !i.length)
        throw new Error("No server id returned");
      return i;
    }
    return {
      storeEntry: async (e, { onprogress: o, signal: s }) => {
        if (!I(e) || !K(e.file))
          return;
        const a = b(e.file), { resume: u, valueKey: l } = v;
        let n = e.state[l], f = 0, r = [];
        if (o(N()), !n)
          n = await F(e.file, { signal: s }, e), E(e, n);
        else
          try {
            const t = await O(n, { signal: s }, e);
            f = t.offset ?? 0, r = t.chunks ?? [];
          } catch (t) {
            throw x(s, t) && E(e, n), t;
          }
        try {
          const t = await A(
            // get chunks that still need to be uploaded
            a.filter((i) => !r.some(
              (d) => d.index === i.index || d.offset === i.offset
            ) && i.offset >= f),
            // need to add them to this transfer
            n,
            // file headers
            {
              uploadName: `${e.file.name}`,
              uploadLength: `${e.file.size}`
            },
            // upload progress
            {
              onprogress: o,
              signal: s
            },
            // entry reference
            e
          );
          r = [...r, ...t];
        } catch (t) {
          throw u && x(s, t) && E(e, n), t;
        }
        return M(n, r, { signal: s }, e);
      },
      releaseEntry: async (e, o, s) => {
        const {
          url: a,
          resolveRequest: { release: u = k }
        } = v, { signal: l } = s ?? {}, n = await u({
          url: a,
          options: {
            method: "DELETE",
            queryString: {
              id: e
            }
          },
          entry: o,
          id: e
        });
        return await S(n.url, {
          ...n.options,
          signal: l
        }), !0;
      }
    };
  }
});
export {
  oe as ChunkedUploadStore
};
