// filepond/esm/extensions/console-view.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createExtension as I } from "./common/createExtension.js";
import { bytesToNaturalFileSize as P } from "../utils/file.js";
import { isDirectoryEntry as D, isNumber as G } from "../utils/test.js";
import { Status as u } from "../common/status.js";
import { clamp as N } from "../utils/math.js";
import { arrayRemoveFalsy as $ } from "../utils/array.js";
import { log as m, clear as R } from "../common/console.js";
import { toSpaceSeparatedString as T } from "../elements/common/string.js";
let S = 0;
const X = I({
  name: "ConsoleView",
  type: "view",
  props: {
    clearBeforeLog: !1,
    debounce: !0
  },
  factory: ({ props: w, extensionName: x }, C) => {
    const { on: b, insertEntries: F, updateEntry: L, removeEntries: B } = C;
    function p(e, r = /* @__PURE__ */ new WeakSet()) {
      if (e === null || typeof e != "object")
        return e;
      if (r.has(e))
        return null;
      if (r.add(e), Array.isArray(e))
        return e.map((o) => p(o, r));
      try {
        return structuredClone(e);
      } catch {
        if (/Error/.test(e.constructor.name))
          return {
            code: e.code,
            message: e.message
          };
        let t = {};
        for (const [l, a] of Object.entries(e))
          t[l] = p(a);
        return t;
      }
    }
    function O(e) {
      return p(e);
    }
    function U(e) {
      if (!G(e))
        return;
      if (e === 1 / 0)
        return "∞ busy";
      const r = N(e), o = 15, t = Math.round(r * o);
      return "█".repeat(t) + "░".repeat(o - t) + " " + Math.round(r * 100) + "%";
    }
    function k(e, r, o) {
      return !!e.status;
    }
    function d(e, r, o, t = "") {
      let l = "inherit";
      const f = $(Object.values(e.extension ?? {})).filter(k).map(({ status: n }) => {
        let E = [], s = l, h = "", c = "";
        return n.type === u.Warning && (c = "▲", s = "Orange"), n.type === u.Error && (c = "✖︎", s = "OrangeRed"), n.type === u.Success && (c = "✔", s = "YellowGreen"), n.type === u.Info && (c = "i", s = "SkyBlue"), n.type === u.System && (n.code.includes("BUSY") ? (c = "⧗", h = ` ${U(n.progress)}`) : n.code.includes("COMPLETE") ? c = "✔" : c = "•", s = "Grey"), E.push({ label: `${c} ${n.code}${h}`, color: s }), E;
      }).flat();
      let i = `%c%s %o	  %c${r}%c${o}%c${t} ${T(
        ...f.map(({ label: n }) => `%c ${n}`)
      )}`;
      i += " ".repeat(Math.max(0, 60 - i.length)) + "", m(
        ...$([
          i,
          "color:grey",
          e.id,
          { "🔍": O(e) },
          "color:grey",
          "color:" + l,
          "color:grey",
          ...f.flat().map(({ color: n }) => `color:${n}`)
        ])
      );
    }
    function v(e, r) {
      d(
        e,
        r,
        e.name ?? e?.state.src ?? "",
        // @ts-ignore
        "size" in e ? ` (${P(e.size)})` : ""
      );
    }
    function A(e, r) {
      d(e, r, e.name + "/");
    }
    function g(e, r, o) {
      e.forEach((t, l) => {
        let a = "";
        const f = l === e.length - 1;
        let i = [...o];
        if (r > 0 && (a = (f ? "└──" : "├──") + " "), D(t))
          return A(t, i.join("") + a), r > 0 && i.push(f ? "    " : "│   "), g(t.entries, r + 1, i);
        v(t, i.join("") + a);
      });
    }
    let y;
    function M(e) {
      const { debounce: r, clearBeforeLog: o } = w, t = () => {
        o && R(), (e.length > 1 || o) && m(`
 handleUpdateEntries(%o)

`, e), g(e, 0, []);
      };
      cancelAnimationFrame(y), r ? y = requestAnimationFrame(t) : t();
    }
    if (window) {
      const e = `$pond${S}`, r = window[e] = {
        insertEntries: F,
        removeEntries: B,
        updateEntry: L
      };
      m(
        `%c${x}: %cFilePond instance available for debugging at %cwindow%c.%c${e}`,
        "color:grey",
        "color:auto",
        "color:grey",
        "color:grey",
        "color:auto",
        { "🔍": r }
      ), S++;
    }
    const z = b("updateEntries", M);
    return {
      destroy: () => {
        z();
      }
    };
  }
});
export {
  X as ConsoleView
};
