// filepond/esm/extensions/data-transfer-loader.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createExtension as A } from "./common/createExtension.js";
import { didAbort as v } from "../utils/abort.js";
import { isFileEntry as y, isDataTransfer as O, isDirectoryEntry as U } from "../utils/test.js";
import { Status as i } from "../common/status.js";
import { shouldLoadWithIdleCallback as F, readEntriesFromDataTransfer as p, dataTransferToFiles as k } from "../common/readEntriesFromDataTransfer.js";
import { flattenTree as w } from "../utils/tree.js";
import { createPerceivedPerformanceProxy as C } from "../common/perceivedPerformanceProxy.js";
const q = A({
  name: "DataTransferLoader",
  type: "loader",
  props: {
    actionLoad: "load",
    actionAbort: "abort",
    mode: "flatten"
  },
  factory: (m, u) => {
    const { props: c, didSetProps: l } = m, {
      on: E,
      removeEntries: D,
      replaceEntry: T,
      pushTask: h,
      abortTasks: S,
      setEntryExtensionStatus: a,
      getEntryExtensionStatus: b
    } = u;
    let o = null;
    l(({ perceivedPerformance: r }) => {
      r === !0 ? o = {
        minDuration: 500,
        maxDuration: 750,
        minStep: 50,
        maxStep: 150
      } : r ? o = r : o = null;
    });
    async function L(r, { signal: n }) {
      const { mode: d } = c;
      a(r, {
        type: i.System,
        code: "LOAD_BUSY",
        progress: 1 / 0
      });
      let s;
      try {
        if (F(r.src)) {
          const x = await (o && !document.hidden ? C(p, {
            ...o
          }) : p)(r.src, {
            signal: n,
            onprogress: ({ loaded: e, total: f }) => {
              a(r, {
                type: i.System,
                code: "LOAD_BUSY",
                progress: e / f,
                values: {
                  processed: e,
                  total: f
                }
              });
            }
          });
          d === "flatten" && (s = w(x, "entries").filter((e) => !U(e)).map((e) => ({
            src: e,
            origin: r.origin,
            containerId: r.id
          })));
        } else
          s = (await k(r.src)).map(
            (t) => ({
              src: t,
              origin: r.origin,
              containerId: r.id
            })
          );
      } catch (t) {
        if (v(n, t)) {
          D(r);
          return;
        }
        throw a(r, {
          type: i.Error,
          code: "LOAD_ERROR",
          values: { error: t }
        }), t;
      }
      a(r, {
        type: i.Success,
        code: "LOAD_COMPLETE"
      }), T(r, s);
    }
    function g(r) {
      if (b(r)?.type === "error" || !y(r) || !O(r.src))
        return;
      const { actionAbort: s } = c;
      if (r.state[s]) {
        S(r.id);
        return;
      }
      h(r.id, L);
    }
    const P = E("updateEntry", g);
    return {
      destroy: () => {
        P();
      }
    };
  }
});
export {
  q as DataTransferLoader
};
