// filepond/esm/extensions/file-input-store.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { debounce as m } from "../utils/debounce.js";
import { createExtension as f } from "./common/createExtension.js";
import { getAsElement as a, setFileInputFilesFromEntries as E } from "../utils/dom.js";
import { isFileEntry as d } from "../utils/test.js";
import { warn as c } from "../common/console.js";
const U = f({
  name: "FileInputStore",
  type: "store",
  props: {
    element: void 0,
    // the event fired on the element when it's updated, defaults to 'update'
    elementUpdateEvent: void 0
  },
  factory: ({ props: r, didSetProps: o }, { on: i }) => {
    let e;
    o(({ element: t }) => {
      e = a(t), e?.type !== "file" && (e = null), t && (!e || e.type !== "file") && c(`FileInputStore: HTMLInputElement not found ${t}`);
    });
    function p(t) {
      if (!e || !t.every(d))
        return;
      const { elementUpdateEvent: u } = r;
      E(
        e,
        // @ts-ignore we know these are file entries
        t.filter((n) => n.extension ? !Object.values(n.extension).some((l) => l.status?.type === "error") : !0),
        {
          customEventType: u
        }
      );
    }
    const s = i("updateEntries", m(p));
    return {
      destroy: () => {
        s();
      }
    };
  }
});
export {
  U as FileInputStore
};
