// filepond/esm/extensions/form-post-store.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createStoreExtension as H } from "./common/createStoreExtension.js";
import { blobToFile as P } from "../utils/file.js";
import { isFileEntry as B, isFile as g } from "../utils/test.js";
import { passthrough as p } from "../utils/placeholder.js";
import { xhr as m, getResponseHeaders as q, getFilenameFromResponseHeaders as w } from "../utils/xhr.js";
const k = H({
  name: "FormPostStore",
  props: {
    url: "",
    name: "entry",
    fetchMetadata: !0,
    resolveRequest: {},
    resolveResponse: {},
    getBasename: () => "Untitled"
  },
  factory: ({ props: c }, F) => {
    const { updateEntry: S } = F;
    function T(e) {
      const t = q(e), s = w(t);
      if (s)
        return s;
    }
    return {
      storeEntry: async (e, { onprogress: t, signal: s }) => {
        const {
          name: o,
          url: r,
          valueKey: n,
          resolveRequest: { store: a = p },
          resolveResponse: { store: d = ({ value: R }) => R }
        } = c;
        if (!B(e) || !g(e.file))
          return;
        if (g(e.file) && e.state[n] != null)
          return e.state[n];
        const l = await a({
          url: r,
          options: {
            method: "POST",
            formData: [[o, e.file, e.file.name]]
          },
          entry: e
        }), u = await m(l.url, {
          ...l.options,
          signal: s,
          onprogress: t
        });
        return d({
          response: u,
          value: u.response,
          entry: e
        });
      },
      restoreEntry: async (e, t, { onprogress: s, signal: o }) => {
        const {
          url: r,
          fetchMetadata: n,
          getBasename: a,
          resolveRequest: {
            metadata: d = p,
            restore: l = p
          },
          resolveResponse: {
            metadata: u = ({ value: i }) => i,
            restore: R = ({ value: i }) => i
          }
        } = c;
        if (n) {
          const i = await d({
            url: r,
            options: {
              method: "HEAD",
              queryString: {
                id: e
              }
            },
            entry: t
          }), f = await m(i.url, {
            ...i.options,
            signal: o
          }), { contentType: y, contentLength: M, lastModified: D } = q(f), x = u({
            response: f,
            value: {
              name: T(f),
              type: y,
              size: parseInt(M, 10),
              lastModified: new Date(D).getTime()
            },
            entry: t
          });
          S(t, x);
        }
        const E = await l({
          url: r,
          options: {
            method: "GET",
            queryString: {
              id: e
            }
          },
          entry: t
        }), v = await m(E.url, {
          ...E.options,
          responseType: "blob",
          signal: o,
          onprogress: s
        }), { response: h } = v, b = P(
          h,
          // use entry name if defined
          t.name || // else read from response headers
          w(q(v)) || // else fall back
          a(t, h)
        );
        return R({
          response: v,
          value: b,
          entry: t
        });
      },
      releaseEntry: async (e, t, s) => {
        const {
          url: o,
          resolveRequest: { release: r = p }
        } = c, { signal: n } = s ?? {}, a = await r({
          url: o,
          options: {
            method: "DELETE",
            queryString: {
              id: e
            }
          },
          entry: t
        });
        return await m(a.url, {
          ...a.options,
          signal: n
        }), !0;
      }
    };
  }
});
export {
  k as FormPostStore
};
