// filepond/esm/extensions/list-count-validator.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { Status as r } from "../common/status.js";
import { createExtension as E } from "./common/createExtension.js";
const d = E({
  name: "ListCountValidator",
  type: "validator",
  props: {
    minFiles: 0,
    maxFiles: 1 / 0
  },
  factory: ({ props: a, didSetProps: l }, { on: m, setExtensionStatus: e }) => {
    let n = !1;
    l(({ minFiles: i, maxFiles: t }) => {
      n = i !== 0 || t !== 1 / 0;
    });
    function u(i) {
      if (!n)
        return;
      const t = i.length, { minFiles: o, maxFiles: s } = a;
      if (t < o)
        return e({
          type: r.Error,
          code: "VALIDATION_LIST_ENTRY_COUNT_UNDERFLOW",
          values: { minFiles: o, minFilesUnit: "unitFiles" }
        });
      if (t > s)
        return e({
          type: r.Error,
          code: "VALIDATION_LIST_ENTRY_COUNT_OVERFLOW",
          values: { maxFiles: s, maxFilesUnit: "unitFiles" }
        });
      e({
        type: r.System,
        code: "VALIDATION_COMPLETE"
      });
    }
    const p = m("updateEntries", u);
    return {
      destroy: () => {
        p();
      }
    };
  }
});
export {
  d as ListCountValidator
};
