// filepond/esm/extensions/simulated-loader.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createProgressEvent as T } from "../utils/xhr.js";
import { isFileEntry as H, isBlobOrFile as G, isString as x, isNumber as J } from "../utils/test.js";
import { createExtension as K } from "./common/createExtension.js";
import { Status as n } from "../common/status.js";
import { sleep as f } from "../utils/sleep.js";
import { log as N } from "../common/console.js";
const et = K({
  name: "SimulatedLoader",
  type: "loader",
  props: {
    actionLoad: "load",
    actionAbort: "abort",
    bitrate: 1024e3,
    tickrate: 250,
    connectionDelay: 250,
    errorDelay: void 0,
    parallel: 4,
    log: !0,
    fetchFile: void 0
  },
  factory: ({ extensionName: O, props: S, didSetProps: R }, _) => {
    const {
      setEntryExtensionStatus: d,
      getEntryExtensionStatus: P,
      createProgressHandler: z,
      removeEntries: k,
      updateEntry: L,
      pushTask: v,
      abortTasks: B,
      on: M
    } = _;
    let A;
    R(({ bitrate: t = 1024e3, tickrate: e = 250 }) => {
      A = t / 8 * (e / 1e3);
    });
    async function g(t) {
      d(t, {
        type: n.System,
        code: "LOAD_QUEUED"
      });
    }
    async function Q(t) {
      const { src: e, size: l = 1024 * 1024 } = t, { log: o, connectionDelay: r, errorDelay: a } = S;
      if (d(t, {
        type: n.System,
        code: "LOAD_BUSY",
        progress: 1 / 0
      }), await f(r), a) {
        await f(a);
        const i = "Simulated error";
        throw d(t, {
          type: n.Error,
          code: "LOAD_ERROR",
          values: { error: i }
        }), o && c(["did throw load info error", t.id]), i;
      }
      o && c(["did load info", t.id]), L(t, {
        name: e.split("/").pop(),
        type: "plain/text",
        size: l
      });
    }
    async function Y(t, { signal: e }) {
      const { src: l, size: o = 1024 * 1024 } = t, {
        log: r,
        actionLoad: a,
        actionAbort: i,
        tickrate: w,
        connectionDelay: y,
        fetchFile: p,
        errorDelay: m
      } = S;
      d(t, {
        type: n.System,
        code: "LOAD_BUSY",
        progress: 1 / 0
      });
      const u = z(t);
      let b, D, I = !1;
      const s = () => {
        I || (I = !0, clearInterval(b), r && c(["did abort load data", t.id]), L(t, {
          state: {
            [a]: !1,
            [i]: !1
          }
        }), queueMicrotask(() => {
          k(t);
        }), D?.(e.reason));
      };
      if (e.addEventListener("abort", s, { once: !0 }), await f(y), e.aborted)
        throw e.reason;
      if (m) {
        if (await f(m), e.aborted)
          throw e.reason;
        const E = "Simulated error";
        throw d(t, {
          type: n.Error,
          code: "LOAD_ERROR",
          values: { error: E }
        }), r && c(["did throw load data error", t.id]), e.removeEventListener("abort", s), E;
      }
      return await new Promise((E, U) => {
        let h = 0;
        D = U, b = setInterval(async () => {
          if (e.aborted) {
            s();
            return;
          }
          if (h = Math.min(o, h + A), u(T(!0, h, o)), h < o)
            return;
          u(T(!0, o, o)), clearInterval(b);
          let F;
          try {
            if (p)
              F = await p(t, { onprogress: u, signal: e });
            else {
              if (await f(0), e.aborted)
                throw e.reason;
              F = new File(["#".repeat(o)], l.split("/").pop(), {
                type: "plain/text"
              });
            }
          } catch (C) {
            e.removeEventListener("abort", s), U(C);
            return;
          }
          if (e.aborted) {
            s();
            return;
          }
          e.removeEventListener("abort", s), L(t, {
            file: F,
            state: {
              load: !1
            },
            extension: {
              [O]: {
                status: {
                  type: n.Success,
                  code: "LOAD_COMPLETE"
                }
              }
            }
          }), r && c(["did load data", t.id]), E();
        }, w);
      });
    }
    function j(t) {
      const e = P(t), l = Object.keys(e).length > 0;
      if (e?.type === "error" || !H(t) || G(t.file))
        return;
      const { src: r, name: a, size: i } = t, { actionLoad: w, actionAbort: y, parallel: p } = S, m = t.state[w], u = t.state[y];
      if (!x(r))
        return;
      if (u) {
        B(t.id), k(t);
        return;
      }
      if (m === !1)
        return;
      if (!(x(a) && J(i))) {
        v(t.id, Q);
        return;
      }
      if (!l) {
        v(t.id, g);
        return;
      }
      v(t.id, Y, { parallel: p });
    }
    function c(t) {
      N("⧗", O, "(", ...t, ")");
    }
    const q = M("updateEntry", j);
    return {
      destroy: () => {
        q();
      }
    };
  }
});
export {
  et as SimulatedLoader
};
