// filepond/esm/elements/FilePondEntryList/index.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { FilePondSvelteComponentElement as e } from "../FilePondSvelteComponent/index.svelte.js";
import { setBooleanAttribute as n } from "../../utils/dom.js";
import { registerShadowRoot as s } from "../common/extendStyles.js";
import i from "./index.svelte.js";
import a from "./index.css.js";
const r = [
  "disabled",
  "assets",
  "locale",
  "template",
  "propResourceMap",
  "drag",
  "dragGrabTimeout",
  "dragDetachMargin",
  "dragSafetyMargin",
  "drop",
  "dropRoot",
  "dropPadding",
  "animations",
  "entryAnimationProps",
  "entryAnimationOriginMap",
  "entryAnimationStaggerInterval",
  "springDefaults",
  "byteUnits",
  "beforeRenderNode"
], o = [
  // update routes from model
  "onSetEntries",
  "onInsertEntry",
  "onRemoveEntry",
  // registered callbacks
  "setSetEntriesCallback",
  "setInsertEntriesCallback",
  "setRemoveEntriesCallback",
  "setUpdateEntryCallback",
  "setGetEntryExtensionStateCallback",
  "setSetEntryExtensionStateCallback",
  "setPushTaskCallback",
  "setAbortTaskCallback"
], p = ["dragentry", "dragentrystart", "dragentryend", "updateplaceholder"];
class y extends e {
  //
  // /** Called when entries are set */
  // onSetEntries: (entries: FilePondEntry[]) => void;
  // /** Called when entry is inserted */
  // onInsertEntry: (entry: FilePondEntry) => void;
  // /** Called when entry is removed */
  // onRemoveEntry: (detail: { entry: FilePondEntry; index: number[] }) => void;
  // /** Set the callback to use when the view updates entries */
  // setSetEntriesCallback: (cb: (entries: FilePondEntry[]) => void) => void;
  // /** Set the callback to use when user wants to remove entry */
  // setInsertEntriesCallback: (
  //     cb: (entry: FilePondEntrySource | FilePondEntrySource[], index?: number | number[]) => void
  // ) => void;
  // /** Set the callback to use when user wants to remove entry */
  // setRemoveEntriesCallback: (
  //     cb: (
  //         ...needles: Needle[]
  //     ) =>
  //         | ({ entry: FilePondEntry; index: number[] } | void)[]
  //         | { entry: FilePondEntry; index: number[] }
  //         | void
  // ) => void;
  // /** Set the callback to use for updating entry state */
  // setUpdateEntryCallback: (cb: (needle: Needle, ...props: any[]) => void) => void;
  // /** Set the callback to use to get the current entry state */
  // setGetEntryExtensionStateCallback: (
  //     cb: (entry: FilePondEntry) => { [key: string]: any }
  // ) => void;
  // /** Set the callback to use to update the entry status */
  // setSetEntryExtensionStateCallback: (
  //     cb: (entry: FilePondEntry, props: { [key: string]: any }) => void
  // ) => void;
  // setPushTaskCallback: (cb: (id: string, fn: Function) => void) => void;
  // setAbortTaskCallback: (cb: (id: string, fn: Function) => void) => void;
  constructor() {
    super(i, {
      properties: r,
      methods: o,
      events: p
    }), s(this._root, a);
  }
  connectedCallback() {
    super.connectedCallback(), this.addListener("updateentries", (t) => {
      n(this, "empty", t.detail === 0);
    });
  }
}
function S() {
  return {
    clipboard: "drop",
    drop: "rise",
    input: "slide",
    remote: "rise"
  };
}
function E() {
  return {
    stiffness: 0.1,
    damping: 0.495,
    precision: 1e-3
  };
}
function O() {
  return {
    // spring state when dragging outside list
    dissolve: {
      scale: 1,
      opacity: 0.5
    },
    // spring state while dragging
    lift: {
      translationSpringOptions: { stiffness: 0.1, damping: 0.4 },
      scaleSpringOptions: { stiffness: 0.1, damping: 0.35 },
      scale: 1.025,
      opacity: 1
    },
    release: {
      scale: 1,
      opacity: 1
    },
    // spring target when removed
    fall: {
      scaleSpringOptions: { stiffness: 0.1, damping: 0.9 },
      scale: 0.95,
      opacitySpringOptions: { stiffness: 0.25, damping: 0.95 },
      opacity: 0
    },
    // fade in from smaller size
    rise: {
      translationFrom: { x: 0, y: -5 },
      scaleSpringOptions: { stiffness: 0.1, damping: 0.31 },
      scaleFrom: 0.95,
      scale: 1,
      opacitySpringOptions: { stiffness: 0.1, damping: 0.6 },
      opacityFrom: 0,
      opacity: 1
    },
    // fade in from bigger size
    drop: {
      scaleSpringOptions: { stiffness: 0.1, damping: 0.31 },
      scaleFrom: 1.05,
      scale: 1,
      opacitySpringOptions: { stiffness: 0.1, damping: 0.6 },
      opacityFrom: 0,
      opacity: 1
    },
    // animating in from the top
    slide: {
      translationSpringOptions: { stiffness: 0.05, damping: 0.5 },
      translationFrom: {
        x: 0,
        y: -15
      },
      scaleSpringOptions: { stiffness: 0.1, damping: 0.31 },
      scaleFrom: 0.95,
      scale: 1,
      opacitySpringOptions: { stiffness: 0.1, damping: 0.6 },
      opacityFrom: 0,
      opacity: 1
    }
  };
}
export {
  r as COMPONENT_PROPS,
  y as FilePondEntryListElement,
  S as getDefaultEntryAnimationOriginMap,
  O as getDefaultEntryAnimationProps,
  E as getDefaultSpringOptions
};
