// filepond/esm/extensions/common/createExtension.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { warn as h } from "../../common/console.js";
import { copyDescriptors as k, isObjectValuesEqual as q } from "../../utils/object.js";
import { isString as w, isObject as C } from "../../utils/test.js";
function N(E) {
  const { name: e, type: i, props: m, factory: x } = E || {}, a = (u) => {
    (!w(e) || !e) && h("Extension name missing or invalid");
    function c(t) {
      return t.extension?.[e] ?? {};
    }
    function f(t, n) {
      u.updateEntry(t, {
        extension: {
          [e]: n
        }
      });
    }
    function p(t, n) {
      f(t, { status: n });
    }
    function s(t) {
      return c(t).status ?? {};
    }
    function d(t) {
      const n = s(t), { type: b, code: P } = n ?? s(t);
      return ({ lengthComputable: j, loaded: O, total: v }) => {
        p(t, {
          type: b,
          code: P,
          progress: j ? O / v : 1 / 0
        });
      };
    }
    let r = { ...m }, l = [];
    function g(t) {
      l.push(t), t(o());
    }
    const y = x(
      // instance
      {
        // reference to current props so extension always has latest values
        props: r,
        // called when props updated (when setProps called and on init)
        didSetProps: g,
        // return name of this extension
        extensionName: e,
        // return type of this extension
        extensionType: i
      },
      {
        // merge extension manager with extension context
        ...u,
        getEntryExtensionState: c,
        setEntryExtensionState: f,
        setEntryExtensionStatus: p,
        getEntryExtensionStatus: s,
        createProgressHandler: d
      }
    );
    function S(t) {
      C(t) && (q(t, r) || (Object.assign(r, t), l.forEach((n) => n(o()))));
    }
    function o() {
      return { ...r };
    }
    return k(y, {
      setProps: S,
      getProps: o,
      get name() {
        return e;
      },
      destroy() {
        y.destroy();
      }
    });
  };
  return Object.defineProperties(a, {
    name: { value: e, writable: !1 },
    type: { value: i, writable: !1 }
  }), a;
}
export {
  N as createExtension
};
