// filepond/esm/extensions/common/createStoreExtension.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createExtension as at } from "./createExtension.js";
import { isNullOrUndefined as f, isFile as h, isFileEntry as rt, isBlob as ct } from "../../utils/test.js";
import { Status as n } from "../../common/status.js";
import { debounce as ut } from "../../utils/debounce.js";
import { createPerceivedPerformanceProxy as it } from "../../common/perceivedPerformanceProxy.js";
import { didAbort as P } from "../../utils/abort.js";
function _t(k) {
  const { name: d, props: B, factory: Y } = k;
  return at({
    name: d,
    type: "store",
    props: {
      perceivedPerformance: null,
      parallel: 4,
      actionStore: "store",
      actionLoad: "load",
      actionAbort: "abort",
      valueKey: "value",
      ...B
    },
    factory: (b, U) => {
      const { props: i, didSetProps: N } = b, {
        on: y,
        setExtensionStatus: I,
        updateEntry: l,
        removeEntries: C,
        pushTask: S,
        abortTask: F,
        getEntryExtensionState: D,
        getEntryExtensionStatus: A,
        setEntryExtensionStatus: E,
        createProgressHandler: L
      } = U;
      let p = null;
      N(({ perceivedPerformance: t }) => {
        t === !0 ? p = {
          minDuration: 1e3,
          maxDuration: 1500,
          minStep: 50,
          maxStep: 250
        } : t ? p = t : p = null;
      });
      const { restoreEntry: O, storeEntry: K, releaseEntry: v } = Y(b, U) ?? {};
      async function M(t, { signal: s }) {
        E(t, {
          type: n.System,
          code: "STORE_BUSY",
          progress: 1 / 0
        });
        try {
          const o = p && !document.hidden ? it(K, {
            ...p,
            total: t.size
          }) : K, { valueKey: a } = i, u = await o(t, {
            onprogress: L(t),
            signal: s
          });
          f(u) || l(t, {
            state: {
              [a]: u
            },
            extension: {
              [d]: {
                status: {
                  type: n.Success,
                  code: "STORE_COMPLETE"
                }
              }
            }
          });
        } catch (e) {
          if (P(s, e)) {
            const { actionStore: o, actionAbort: a, valueKey: u } = i;
            l(t, {
              state: {
                // don't abort again
                [a]: !1,
                // need to halt store action or will keep storing
                [o]: null,
                // reset storage key
                [u]: null
              },
              extension: {
                [d]: {
                  status: {
                    type: n.System,
                    code: "STORE_ABORT"
                  }
                }
              }
            });
            return;
          }
          throw E(t, {
            type: n.Error,
            code: "STORE_ERROR",
            values: { error: e }
          }), e;
        }
      }
      async function Q(t, { signal: s }) {
        if (!O)
          return;
        E(t, {
          type: n.System,
          code: "RESTORE_BUSY",
          progress: 1 / 0
        });
        const { valueKey: e } = i, o = t.state[e];
        try {
          let a = await O(o, t, {
            onprogress: L(t),
            signal: s
          });
          l(t, {
            // if response is a blob we need BlobLoader to load the source
            src: ct(a) ? a : t.src,
            // if response is a file we can skip straight to file
            file: h(a) ? a : null,
            // new state
            state: {
              // remember storage key
              [e]: o
            },
            extension: {
              [d]: {
                // need to re-evaluate if we can store this file
                canStore: !0,
                // done!
                status: {
                  type: n.System,
                  code: "RESTORE_COMPLETE"
                }
              }
            }
          });
        } catch (a) {
          if (P(s, a)) {
            E(t, {
              type: n.System,
              code: "RESTORE_ABORT"
            });
            return;
          }
          throw E(t, {
            type: n.Error,
            code: "RESTORE_ERROR",
            values: { error: a }
          }), a;
        }
      }
      async function w(t, { signal: s }) {
        const { valueKey: e, actionLoad: o, actionStore: a, shouldStore: u } = i;
        try {
          const r = t.state[e];
          if (f(r))
            return;
          E(t, {
            type: n.System,
            code: "RELEASE_BUSY",
            progress: 1 / 0
          });
          let c = !1;
          if (u) {
            const R = t.state[a] === !1;
            c = await u(t) && R;
          }
          if (O && !h(t.file) && !c) {
            let R = await O(r, t, {
              onprogress: L(t),
              signal: s
            });
            l(t, {
              file: R
            });
          }
          v ? await v(r, t, {
            signal: s
          }) !== !1 && l(t, {
            state: {
              [e]: null,
              [o]: null
            },
            extension: {
              [d]: {
                canStore: !0,
                status: {
                  type: n.System,
                  code: "RELEASE_COMPLETE"
                }
              }
            }
          }) : l(t, {
            state: {
              [e]: null,
              [o]: null
            },
            extension: {
              [d]: {
                status: {
                  type: n.System,
                  code: "RELEASE_COMPLETE"
                }
              }
            }
          }), c && C(t);
        } catch (r) {
          if (P(s, r)) {
            E(t, {
              type: n.System,
              code: "RELEASE_ABORT"
            });
            return;
          }
          throw E(t, {
            type: n.Error,
            code: "RELEASE_ERROR",
            values: { error: r }
          }), r;
        }
      }
      async function z(t) {
        const { valueKey: s } = i, { entry: e } = t;
        if (!v)
          return;
        const o = e.state[s], a = !f(o), r = A(e)?.code !== "RESTORE_ERROR";
        if (!(!a || !r))
          try {
            await v(o, e);
          } catch {
          }
      }
      async function G(t) {
        const s = A(t), { valueKey: e, actionStore: o, actionLoad: a, actionAbort: u } = i, r = t.state[e] ?? null, c = f(t.state[o]) ? null : t.state[o];
        if (f(r)) {
          l(t, {
            state: {
              [e]: r,
              [o]: c
            },
            extension: {
              [d]: {
                // so we can match on extension actions
                actions: [o, a, u],
                // null means undetermined if we can activate, will trigger new test
                canStore: null,
                // awaiting new test
                status: {
                  type: n.System,
                  code: "STORE_LIMBO"
                }
              }
            }
          });
          return;
        }
        if (s.code !== "RESTORE_COMPLETE") {
          if (c !== !1 && s.code === "STORE_COMPLETE") {
            S(t.id, w), E(t, {
              type: n.System,
              code: "STORE_IDLE"
            });
            return;
          }
          l(t, {
            state: {
              [e]: r,
              [o]: c
            },
            extension: {
              [d]: {
                canStore: !0,
                status: {
                  type: n.Success,
                  code: "STORE_COMPLETE"
                }
              }
            }
          });
        }
      }
      async function H(t) {
        const { valueKey: s } = i, e = t.state[s], { canStore: o } = D(t);
        e && (o || E(t, {
          type: n.Success,
          code: "STORE_COMPLETE"
        }));
      }
      async function V(t) {
        const { valueKey: s } = i, e = t.state[s], o = rt(t) && h(t.file);
        l(t, {
          state: {
            [s]: e ?? null
          },
          extension: {
            [d]: {
              canStore: o,
              status: {
                type: n.System,
                code: o ? "STORE_READY" : "STORE_IDLE"
              }
            }
          }
        });
      }
      async function W(t) {
        const { actionStore: s, shouldStore: e } = i;
        if (!e)
          return;
        const o = await e(t);
        l(t, {
          state: {
            [s]: o
          }
        });
      }
      async function j(t) {
        E(t, {
          code: "STORE_QUEUED",
          type: n.System,
          progress: 1 / 0
        });
      }
      async function q(t) {
        const { valueKey: s, parallel: e, actionLoad: o, actionStore: a, actionAbort: u, shouldStore: r } = i, c = t.state[a], R = t.state[o], g = t.state[u], et = t.state[s], x = A(t), { canStore: m = null } = D(t), _ = x?.code === "STORE_BUSY", T = !f(et) && !_;
        if (m === null) {
          S(t.id, V);
          return;
        }
        if (c === null && r) {
          if (g === !1) {
            C(t);
            return;
          }
          S(t.id, W);
          return;
        }
        const ot = !f(O);
        if (T && !h(t.file) && ot && R === !0) {
          S(t.id, Q, { parallel: e });
          return;
        }
        if (T && c === !1) {
          S(t.id, w);
          return;
        }
        if (m === !1) {
          S(t.id, H);
          return;
        }
        if (!T && m && (c === !0 && g === !0)) {
          F(t.id, M), _ || l(t, {
            state: {
              [u]: !1,
              [a]: null
            },
            extension: {
              [d]: {
                status: {
                  type: n.System,
                  code: "STORE_READY"
                }
              }
            }
          });
          return;
        }
        const st = x?.code === "STORE_QUEUED", nt = x?.code === "STORE_ERROR";
        if (!T && !st && !nt && !_ && c === !0 && m) {
          S(t.id, j);
          return;
        }
        if (!T && !_ && c === !0 && m) {
          S(t.id, M, {
            parallel: e
          });
          return;
        }
      }
      function J(t) {
        const { valueKey: s } = i;
        if (s) {
          for (const e of t)
            if (f(e.state[s]))
              return I({
                type: n.Error,
                code: "STORE_AWAITING_COMPLETION",
                meta: { flag: "customError" }
              });
          I({
            type: n.System,
            code: "VALIDATION_COMPLETE"
          });
        }
      }
      const X = y("updateEntryData", G), Z = y("updateEntry", q), $ = y("removeEntry", z), tt = y("updateEntries", ut(J));
      return {
        destroy() {
          X(), Z(), tt(), $();
        }
      };
    }
  });
}
export {
  _t as createStoreExtension
};
