// filepond/esm/utils/test.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { hasOwnProp as f } from "./object.js";
function e(t, n = !0) {
  let r = null;
  return () => n && !y() ? !1 : (r === null && (r = t()), r);
}
function g(t) {
  return t === void 0;
}
function m(t) {
  return t == null;
}
function o(t) {
  return typeof t == "string";
}
function b(t) {
  return typeof t == "number";
}
function F(t) {
  return typeof t == "boolean";
}
function A(t) {
  return typeof t == "function";
}
function O(t) {
  return t instanceof HTMLElement;
}
function E(t) {
  let n;
  try {
    n = new URL(t);
  } catch {
    return !1;
  }
  return n.protocol === "http:" || n.protocol === "https:";
}
function h(t) {
  return t instanceof RegExp;
}
function s(t) {
  return Object.getPrototypeOf(t) === Object.prototype;
}
function D(t) {
  return t && typeof t == "object" && !Array.isArray(t);
}
function w(t) {
  return t && typeof t == "object";
}
function u(t) {
  return Array.isArray(t);
}
function L(t) {
  return o(t) && /data:/.test(t);
}
function T(t) {
  return t instanceof HTMLCanvasElement;
}
function c(t) {
  return t instanceof Blob;
}
function j(t) {
  return t instanceof Blob && !a(t);
}
function a(t) {
  return t instanceof File;
}
function l(t) {
  return t instanceof DataTransfer;
}
function B(t) {
  return !!(c(t) && /video/i.test(t.type));
}
function x(t) {
  return !!(c(t) && /image/i.test(t.type));
}
function P(t) {
  return l(t.src);
}
function R(t) {
  return o(t) && t.startsWith("unit");
}
function U(t) {
  return f(t, "template");
}
function S(t) {
  return !!(t && s(t) && !u(t.entries));
}
function M(t) {
  return !!(t && s(t) && u(t.entries));
}
function C(t) {
  return t?.isDirectory;
}
function H(t) {
  return t?.isFile;
}
function y() {
  return i === null && (i = typeof window < "u" && typeof window.document < "u"), i;
}
let i = null;
const I = e(() => /^((?!chrome|android).)*(safari|iphone|ipad)/i.test(navigator.userAgent)), N = e(() => /Firefox/.test(navigator.userAgent)), K = e(() => /iPhone|iPad|iPod/.test(navigator.userAgent) || p() && navigator.maxTouchPoints >= 1), p = e(() => {
  const { platform: t } = "userAgentData" in navigator ? navigator.userAgentData : navigator;
  return /^mac/i.test(t);
});
export {
  e as createTest,
  u as isArray,
  j as isBlob,
  c as isBlobOrFile,
  F as isBoolean,
  y as isBrowser,
  T as isCanvas,
  l as isDataTransfer,
  P as isDataTransferEntry,
  L as isDataURL,
  M as isDirectoryEntry,
  O as isElement,
  a as isFile,
  S as isFileEntry,
  C as isFileSystemDirectoryEntry,
  H as isFileSystemFileEntry,
  N as isFirefox,
  A as isFunction,
  K as isIOS,
  x as isImageFile,
  U as isLocaleTemplate,
  R as isLocaleUnitKey,
  p as isMac,
  m as isNullOrUndefined,
  b as isNumber,
  D as isObject,
  s as isObjectLiteral,
  w as isObjectOrArray,
  h as isRegExp,
  I as isSafari,
  o as isString,
  E as isURL,
  g as isUndefined,
  B as isVideoFile
};
