// filepond/esm/utils/xhr.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { toKebabCase as C, toCamelCase as L } from "./string.js";
import { toURL as P } from "./url.js";
import { noop as F } from "./placeholder.js";
import { thread as U, createThreadWorker as $ } from "./thread.js";
import { arrayRemoveFalsy as y } from "./array.js";
import { httpRequest as h } from "../workers/httpRequest.js";
function T(t, r) {
  const {
    method: e,
    queryString: n,
    headers: s,
    data: a,
    formData: m,
    responseType: g,
    withCredentials: R,
    timeout: w,
    useWebWorkers: H = !1,
    workersURL: q,
    signal: p,
    onprogress: v = F
  } = r ?? {}, l = {
    url: E(t, n),
    responseType: g,
    method: e,
    headers: Object.entries(s ?? {}).map(([i, u]) => [
      C(i),
      `${u}`
    ]),
    data: a,
    formData: m,
    timeout: w,
    withCredentials: R
  }, f = { onprogress: v, signal: p };
  function d(i) {
    return new Promise((u, o) => {
      i.then((c) => {
        u({
          getAllResponseHeaders: () => c.responseHeaders,
          response: c.response
        });
      }).catch(o);
    });
  }
  return d(
    H ? (
      // @ts-ignore fix types
      U($(q, h), [l], f)
    ) : new Promise(
      (i, u) => (
        // httpRequest()
        h(
          l,
          (o, c) => {
            if (o !== null) {
              u(o === p?.reason ? o : new Error(`${o}`));
              return;
            }
            i(c);
          },
          f
        )
      )
    )
  );
}
function E(t, r = {}) {
  const e = P(t);
  return Object.entries(r).forEach(
    ([n, s]) => e.searchParams.append(n, `${s}`)
  ), `${e}`;
}
function j(t, r, e) {
  return new ProgressEvent("progress", {
    lengthComputable: t || !1,
    loaded: t ? r && e === 1 ? r * 100 : r : 0,
    total: t ? e === 1 ? 100 : e : 0
  });
}
function I(t) {
  return t ? y(
    t.getAllResponseHeaders().split(`
`).map((e) => {
      const n = e.match(/(^.*?):/) || [], [s, a] = n;
      if (!s)
        return;
      const m = e.replace(s, "").trim();
      return [a, m];
    })
  ).reduce(
    (e, n) => {
      const [s, a] = n;
      return e[L(s)] = a, e;
    },
    {}
  ) : {};
}
function K(t) {
  const { contentDisposition: r } = t;
  return !r || !r.length ? null : W(r);
}
function W(t) {
  if (!t.toLowerCase().startsWith("attachment"))
    return null;
  const r = t.split(/filename=|filename\*=.+''/i).splice(1).map((e) => e.trim().replace(/^["']|[;"']{0,2}$/g, "")).filter((e) => e.length);
  return r.length ? decodeURI(r[r.length - 1]) : null;
}
function S(t, r) {
  const e = r.toLowerCase().split(`
`).find((n) => n.includes(t));
  return e ? e.split(":")[1].trim() : void 0;
}
export {
  j as createProgressEvent,
  W as getFilenameFromContentDispositionHeader,
  K as getFilenameFromResponseHeaders,
  S as getResponseHeaderValue,
  I as getResponseHeaders,
  T as xhr
};
