// filepond/esm/elements/FilePondDefault/createFilePondExtensionSet.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { isObject as c, isArray as E } from "../../utils/test.js";
import { arrayInsertAtIndex as T } from "../../utils/array.js";
import { warn as S } from "../../common/console.js";
import { FileInputSource as w } from "../../extensions/file-input-source.js";
import { DataTransferLoader as F } from "../../extensions/data-transfer-loader.js";
import { ValueCallbackStore as b } from "../../extensions/value-callback-store.js";
import { FileExtensionValidator as g } from "../../extensions/file-extension-validator.js";
import { FileMimeTypeValidator as v } from "../../extensions/file-mime-type-validator.js";
import { EntryListView as V } from "../../extensions/entry-list-view.js";
import { hasOwnProp as L } from "../../utils/object.js";
const O = [
  "source",
  "loader",
  "validator",
  "transform",
  "resource",
  "view",
  "store"
];
function h(e) {
  return O.includes(e);
}
function y(e) {
  return E(e) ? e[0] : c(e) && L(e, "insert") ? e.insert : e;
}
function m(e) {
  return y(e).name;
}
function u(e) {
  return y(e).type;
}
function A(e) {
  if (!c(e))
    return;
  const { insert: r, options: t, ...f } = e;
  return f;
}
const d = { name: "Transform", type: "transform" }, x = { name: "Resource", type: "resource" };
function M(e = []) {
  let r = [
    w,
    F,
    g,
    v,
    // the default extension set doesn't have a Transform extension, so we create a slot so we can auto insert transform extensions there, the slot is removed when we return the extension set
    d,
    // same for Resource extensions, these should run before views so views can use their generated entry state
    x,
    V,
    b
  ];
  for (const t of e) {
    let f = m(t), l = u(t), i = r.findIndex((n) => m(n) === f);
    if (i > -1) {
      r[i] = t;
      continue;
    }
    let o, p = 1, s = A(t), a = !1;
    if (s ? (a = !!s.before, p = a ? 0 : 1, o = s.before || s.after) : o = l, i = s && !h(o) ? r.findIndex((n) => m(n) === o) : a ? r.findIndex((n) => u(n) === o) : r.findLastIndex(
      (n) => u(n) === o
    ), i === -1) {
      S(`No valid insertion index found for extension "${f}" with type "${l}"`);
      continue;
    }
    const I = c(t) ? (
      //  @ts-ignore
      [t.insert, t.options]
    ) : t;
    r = T(r, i + p, I);
  }
  return r.filter(
    (t) => t !== d && t !== x
  );
}
export {
  M as createFilePondExtensionSet
};
