// filepond/esm/elements/FilePondDefault/index.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { FilePondInputElement as b } from "../FilePondInput/index.js";
import { FilePondEntryListElement as g, getDefaultSpringOptions as c, getDefaultEntryAnimationOriginMap as y, getDefaultEntryAnimationProps as A } from "../FilePondEntryList/index.js";
import { FilePondDropAreaElement as E } from "../FilePondDropArea/index.js";
import { FilePondDropIndicatorElement as D } from "../FilePondDropIndicator/index.js";
import { hasDefinedTag as P, defineCustomElements as w, defineCustomElement as L, setBooleanAttribute as a, h as d, addListener as s, dispatchCustomEvent as h } from "../../utils/dom.js";
import { isString as u } from "../../utils/test.js";
import { assets as x } from "../../assets/index.js";
import C from "./index.css.js";
import { createFilePondEntryList as O } from "../../templates/entry.js";
import { createFilePondExtensionSet as k } from "./createFilePondExtensionSet.js";
const _ = ["springDefaults", "animations"];
let l;
class v extends b {
  // Child components
  #t = {};
  /** Holds reference to attribution link element */
  #e;
  /** Holds references to event subscriptions so we can more easily unsub */
  #r = [];
  /** Automatically passes value to child elements, for usage see `FilePondSvelteComponentElement` */
  set springDefaults(t) {
    Object.values(this.#t).forEach((e) => {
      e.springDefaults = t;
    });
  }
  /** Automatically passes value to child elements, for usage see `FilePondSvelteComponentElement` */
  set animations(t) {
    Object.values(this.#t).forEach((e) => {
      e.animations = t;
    });
  }
  /** Wraps `createFilePondExtensionSet` so we always set the default extension set */
  set extensions(t) {
    super.extensions = k(t);
  }
  /** Set to `true` to remove drop area */
  set noDrop(t) {
    a(this, "nodrop", t), t ? (this.#t.dropArea.remove(), this.#t.dropIndicator.remove(), Object.assign(this, {
      EntryListView: {
        drop: !1
      }
    })) : (this._root.prepend(this.#t.dropArea, this.#t.dropIndicator), Object.assign(this, {
      EntryListView: {
        drop: !0
      }
    })), this.setBrowseButtonLabelKey(t ? "browse" : "browseAndDrop");
  }
  /** Returns current nodrop state */
  get noDrop() {
    return this.hasAttribute("nodrop");
  }
  /** 
      A programmatic way to toggle the attribution link on/off.
  
      When set to `true` this property automatically adds the `noattribution` attribute to the `<file-pond>` element.
  
      ```js
      const element = document.querySelector('file-pond');
      element.noAttribution = true;
      ```
      */
  set noAttribution(t) {
    t ? (a(this, "noattribution", !0), this.#e.remove()) : (a(this, "noattribution", !1), this._root.append(this.#e));
  }
  /** Returns current noattribution state */
  get noAttribution() {
    return !this.#e?.parentNode;
  }
  static get observedAttributes() {
    return [...super.observedAttributes, "noattribution", "nodrop"];
  }
  attributeChangedCallback(t, e, r) {
    if (t === "noattribution") {
      this.noAttribution = u(r);
      return;
    }
    if (t === "nodrop") {
      this.noDrop = u(r);
      return;
    }
    super.attributeChangedCallback(t, e, r);
  }
  constructor() {
    super({
      styles: [C]
    }), this.setBrowseButtonLabelKey("browseAndDrop");
    const t = d("file-pond-entry-list", {
      part: "entry-list"
    }), e = d("file-pond-drop-area", {
      part: "drop-area"
    }), r = d("file-pond-drop-indicator", {
      part: "drop-indicator"
    });
    this.#t = {
      entryList: t,
      dropArea: e,
      dropIndicator: r
    };
    const i = /* @__PURE__ */ new Set(["dragging", "virtualized", "selected", "checked"]);
    function n(o) {
      if (!o || i.has(o))
        return;
      const m = Array.from(i.add(o)).join(",");
      t.setAttribute("exportparts", m.replace(/ /g, ","));
    }
    const f = l?.EntryListView?.template || O();
    Object.assign(this, {
      // add items view
      extensions: this.extensions,
      // show progress indicator for data transfers
      DataTransferLoader: {
        perceivedPerformance: !0
      },
      // set up items view extension
      EntryListView: {
        // the element that the item list will be appended to
        element: this.#t.entryList,
        // the root element to use for dragging and dropping components, defaults to the list itself
        dropRoot: this.#t.dropArea,
        // assets to use
        assets: x,
        // the nodes to render
        template: f,
        // called before rendering a node, allows dynamically modifying a node or adding nodes
        beforeRenderNode(o) {
          return n(o.props?.part || o.attrs?.part), o;
        },
        // animations
        entryAnimationProps: A(),
        entryAnimationOriginMap: y(),
        springDefaults: c()
      }
    }), this.springDefaults = c(), _.forEach((o) => {
      this[o] = t[o];
    }), this.#e = F({
      caption: "Powered by FilePond"
    }), Object.assign(this, l);
  }
  connectedCallback() {
    super.connectedCallback();
    const { dropArea: t, dropIndicator: e, entryList: r } = this.#t;
    this.hasAttribute("nodrop") || this._root.prepend(t, e), this._root.append(r), this.hasAttribute("noattribution") || this._root.append(this.#e), this.#r.push(
      // route clicks on drop area to browse button
      s(t, "click", () => {
        this.browse();
      }),
      // did compute target rect
      s(t, "computerect", (i) => {
        if (!i.detail)
          return;
        const n = i.detail;
        h(this, "computerect", { detail: n });
      }),
      // did update visual rect
      s(t, "updaterect", (i) => {
        if (!i.detail)
          return;
        const n = i.detail;
        this._root.style.setProperty("--width", n.width), this._root.style.setProperty("--height", n.height), h(this, "updaterect", { detail: n });
      }),
      // link up placeholder position with drop indicator
      s(r, "updateplaceholder", (i) => {
        e.indicatorRect = i.detail;
      }),
      // these two listeners toggle the dragging attribute to the file-pond element, we do this so we can move the file-pond element that is being interacted with to the front, so the dragged item also renders on top. Additionally they prevent interaction with slot content and attribution link while dragging
      s(r, "dragentrystart", () => {
        a(this, "dragging", !0), this._slot.inert = !0, this.#e.inert = !0;
      }),
      s(r, "dragentryend", () => {
        a(this, "dragging", !1), this._slot.inert = !1, this.#e.inert = !1;
      })
    );
  }
  /** Called each time the element is removed from the document. */
  disconnectedCallback() {
    super.disconnectedCallback(), Object.values(this.#t).forEach((t) => t.remove()), this.#e.remove(), this.#r.forEach((t) => t()), this.#r = [];
  }
}
function F(p) {
  const { caption: t = "" } = p || {};
  return d("a", {
    textContent: t,
    href: "https://filepond.com",
    target: "_tab",
    rel: "noopener noreferrer nofollow",
    part: "attribution-link",
    tabindex: "-1"
  });
}
function N(p) {
  const t = "file-pond";
  return l = p, P(t) || (w({
    [`${t}-entry-list`]: g,
    [`${t}-drop-area`]: E,
    [`${t}-drop-indicator`]: D
  }), L(t, v)), Array.from(document.querySelectorAll(t));
}
export {
  v as FilePondElement,
  N as defineFilePond
};
