// filepond/esm/core/extensionManager.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { createTaskScheduler as z } from "./taskScheduler.js";
import { pubsub as D } from "../utils/pubsub.js";
import { isArray as T } from "../utils/test.js";
import { arrayItemsEqual as G, arrayRemoveInPlace as H, arraySortByItemProp as J } from "../utils/array.js";
function X(k) {
  const { entryTree: a } = k, { on: I, pub: m } = D(), c = [], { insertEntries: A, replaceEntry: v, updateEntry: j, removeEntries: C } = a, f = z({ log: void 0 }), x = {
    extension: {}
  };
  function l(n, r) {
    x.extension[n] = {
      ...g(n),
      ...r
    }, m("updateExtensionState", x.extension);
  }
  function g(n) {
    return x.extension[n] ?? {};
  }
  function O(n, r) {
    l(n, { status: r });
  }
  function q(n) {
    return x.extension[n].status;
  }
  a.on("removeEntry", ({ entry: n }) => {
    f.abortTasks(n.id);
  });
  function M(n) {
    const r = (e, i) => a.on(e, i), o = [], t = { current: void 0 };
    t.current = n({
      // events
      on: r,
      // get root extension state
      setExtensionState: s((e) => {
        l(t.current.name, e);
      }),
      getExtensionState: s(() => g(t.current.name)),
      // get root extension status
      setExtensionStatus: s(
        (e) => O(t.current.name, e)
      ),
      // @ts-ignore
      getExtensionStatus: s(() => q(t.current.name)),
      pushTask: function(e, i, u) {
        const R = E(
          t.current.name
        );
        a.findEntries(e) && f.pushTask(e, i, {
          // set order to factor of 100 so in theory there's plenty room for adding manual order
          order: R.index * 100,
          // when params is a function it is called on run task
          params: () => [a.findEntries(e)],
          // add custom options
          ...u
        });
      },
      abortTask: function(e, i) {
        f.abortTask(e, i);
      },
      abortTasks: (e) => {
        f.abortTasks(e);
      },
      setEntries: function(e) {
        a.entries = e;
      },
      getEntries: () => a.entries,
      // manipulating entry list
      insertEntries: A,
      removeEntries: C,
      replaceEntry: v,
      updateEntry: j
    });
    function s(e) {
      return (...i) => {
        if (t.current) {
          e(...i);
          return;
        }
        o.push([e, i]);
      };
    }
    return o.forEach(([e, i]) => {
      e(...i);
    }), t.current;
  }
  function S() {
    f.abortTasks(), c.map((n) => n.instance).forEach((n) => n.destroy()), c.length = 0;
  }
  const h = {
    // subscribe to events
    on: I,
    get extensions() {
      return c.map(({ factory: n }) => n);
    },
    set extensions(n) {
      if (!T(n))
        return;
      const r = n.map(
        (t) => Array.isArray(t) ? t[0] : t
      );
      if (G(r, h.extensions)) {
        for (const t of n) {
          if (!T(t))
            continue;
          const s = c.find((e) => t[0] === e.factory);
          d(s, t[1]);
        }
        return;
      }
      n.length === 0 && (p.clear(), S());
      for (const t of c)
        if (!r.includes(t.factory)) {
          t.instance.destroy(), H(c, (s) => s === t);
          continue;
        }
      for (const [t, s] of Object.entries(r)) {
        const e = parseInt(t, 10), i = c.find((u) => s === u.factory);
        if (i) {
          if (i.index = e, Array.isArray(n[e])) {
            const u = n[e][1];
            d(i, u);
          }
          continue;
        }
        if (c.push({
          index: e,
          factory: s,
          instance: M(s)
        }), Array.isArray(n[e])) {
          const u = n[e][1];
          d(c.at(-1), u);
        }
      }
      J(c, "index"), Object.entries(p.get("*") ?? []).forEach(
        ([t, s]) => {
          P(t, s);
        }
      );
      for (const [t, s] of p)
        t !== "*" && b(t, s);
      const o = c.map((t) => t.instance.name);
      m("setExtensions", { extensionNames: o }), a.entries = [...a.entries];
    },
    propagateExtensionProperty: P,
    setExtensionProperties: b,
    getExtensionProperties: W,
    // access manager state
    getState() {
      return x.extension;
    },
    // destroy FilePond instance
    destroy() {
      S();
    }
  };
  function P(n, r) {
    const o = { [n]: r }, t = B(n);
    for (const s of t)
      d(s, o);
    y("*", o);
  }
  function b(n, r) {
    const o = E(n);
    o ? d(o, r) : y(n, r);
  }
  function W(n) {
    const r = E(n);
    if (!r)
      return;
    const { instance: o } = r;
    if (o)
      return o.getProps();
  }
  function d(n, r) {
    r && (n.instance.setProps(r), y(n.instance.name, r));
  }
  function B(n) {
    return c.filter((r) => {
      const o = r.instance.getProps();
      return Object.keys(o).includes(n);
    });
  }
  function E(n) {
    return c.find((r) => r.instance.name === n);
  }
  const p = /* @__PURE__ */ new Map();
  function y(n, r) {
    p.set(n, {
      ...p.get(n),
      ...r
    });
  }
  return h;
}
export {
  X as createExtensionManager
};
