// filepond/esm/core/taskScheduler.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { arrayRemoveInPlace as k } from "../utils/array.js";
import { didAbort as z } from "../utils/abort.js";
import { pubsub as J } from "../utils/pubsub.js";
import { isString as K, isFunction as h } from "../utils/test.js";
const o = {
  QUEUED: 1,
  ACTIVE: 2,
  FAILED: 3,
  HALTED: 4
};
function _(v) {
  const { log: g = void 0 } = v ?? {}, { on: C, pub: f } = J(), s = [], i = /* @__PURE__ */ new Map();
  function F(t) {
    return t.state === o.QUEUED || t.state === o.ACTIVE;
  }
  function m(t) {
    return t.state === o.QUEUED;
  }
  function Q(t) {
    return t.state === o.ACTIVE;
  }
  function L(t, r) {
    return K(t) && h(r);
  }
  function E(t, r) {
    let n = 0;
    for (let e = 0; e < r.length; e++)
      t.order >= r[e].order && (n = e + 1);
    r.splice(n, 0, t);
  }
  function B(t) {
    E(t, s);
    let r = i.get(t.group);
    r || (r = [], i.set(t.group, r)), E(t, r);
  }
  function d(t) {
    k(
      s,
      (n) => n.group === t.group && n.fn === t.fn
    );
    const r = i.get(t.group);
    r && k(r, (n) => n.fn === t.fn);
  }
  function H(t) {
    k(s, (r) => r.group === t), i.delete(t);
  }
  function c(t) {
    return i.get(t) ?? [];
  }
  function A(t, r) {
    return c(t)[0].state === r;
  }
  function q(t) {
    return c(t).filter(F).length > 0;
  }
  function V(t) {
    const r = c(t);
    for (const n of r)
      U(n);
  }
  function b(t) {
    return c(t).filter(m)[0];
  }
  function S(t, r) {
    const n = c(t);
    if (n)
      return n.find((e) => e.fn === r);
  }
  function I(t, r, n) {
    const { isSoftFailure: e = !1 } = n || {}, l = i.get(t) ?? [];
    for (const a of l)
      e && a.ignoreSoftFailure || (a.state = r);
  }
  function w() {
    return Array.from(new Set(s.map((t) => t.group)));
  }
  function N() {
    return i.keys();
  }
  function R() {
    return s.filter(m).length > 0;
  }
  function P() {
    return s.filter(Q);
  }
  function M(t) {
    return P().filter((r) => r.fn === t.fn);
  }
  function y(t) {
    return t ? t.parallel === 1 / 0 || M(t).length < t.parallel : !1;
  }
  function p() {
    queueMicrotask(D);
  }
  function D() {
    if (!R()) {
      g?.(s), f("idle");
      return;
    }
    g?.(s);
    const t = w();
    for (const r of t) {
      if (!A(r, o.QUEUED)) {
        if (A(r, o.HALTED)) {
          const e = b(r);
          if (!e || e.ignoreSoftFailure === !1 || !y(e))
            continue;
          G(e);
        }
        continue;
      }
      const n = b(r);
      y(n) && G(n);
    }
  }
  async function G(t) {
    const { group: r, fn: n, params: e, abortController: l } = t, { signal: a } = l;
    t.state = o.ACTIVE;
    try {
      const u = h(e) ? e() : e, T = await n(...u, { signal: a });
      d(t), T === !1 && I(r, o.HALTED, { isSoftFailure: !0 });
    } catch (u) {
      if (z(a, u))
        return;
      f("error", u), I(r, o.HALTED), t.state = o.FAILED;
    }
    q(r) || f("complete", r), p();
  }
  function U(t) {
    t.abortController.signal.aborted || t.abortController.abort();
  }
  function W(t, r, n) {
    const {
      parallel: e = 1 / 0,
      order: l = 0,
      params: a = [],
      ignoreSoftFailure: u = !1
    } = n ?? {};
    if (!L(t, r))
      return;
    const T = S(t, r);
    if (T) {
      T?.state === o.FAILED && (T.state = o.QUEUED, D());
      return;
    }
    B({
      group: t,
      fn: r,
      params: a,
      order: l,
      parallel: e,
      ignoreSoftFailure: u,
      state: o.QUEUED,
      abortController: new AbortController()
    }), p();
  }
  function j(t, r) {
    const n = S(t, r);
    n && (U(n), d(n), p());
  }
  function x(t) {
    if (!t) {
      for (const r of N())
        x(r);
      return;
    }
    V(t), H(t), f("abort", t), f("complete", t), p();
  }
  return {
    on: C,
    pushTask: W,
    abortTask: j,
    abortTasks: x
  };
}
export {
  _ as createTaskScheduler
};
