// filepond/esm/svelte/svelte/src/internal/client/dom/elements/attributes.js
import { LOADING_ATTR_SYMBOL as j, ATTRIBUTES_CACHE as B, IS_XHTML as I } from "../../constants.js";
import { NAMESPACE_HTML as V, ATTACHMENT_KEY as Y, UNINITIALIZED as D } from "../../../../constants.js";
import { get as z, set_active_reaction as M, set_active_effect as k, active_reaction as K, active_effect as X } from "../../runtime.js";
import { get_prototype_of as Z, get_descriptors as q } from "../../../shared/utils.js";
import { managed as F, destroy_effect as R, branch as J, effect as Q } from "../../reactivity/effects.js";
import { flatten as W } from "../../reactivity/async.js";
import { delegated as x, delegate as tt, create_event as et } from "./events.js";
import { autofocus as it } from "./misc.js";
import { is_capture_event as rt, normalize_attribute as st, can_delegate_event as ot } from "../../../../utils.js";
import { attach as ft } from "./attachments.js";
import { clsx as at } from "../../../shared/attributes.js";
import { set_class as ct } from "./class.js";
import { set_style as ut } from "./style.js";
import { select_option as P, init_select as lt } from "./bindings/select.js";
const T = /* @__PURE__ */ Symbol("class"), g = /* @__PURE__ */ Symbol("style"), G = /* @__PURE__ */ Symbol("is custom element"), H = /* @__PURE__ */ Symbol("is html"), nt = I ? "option" : "OPTION", _t = I ? "select" : "SELECT", vt = I ? "progress" : "PROGRESS";
function wt(t, i) {
  var e = m(t);
  e.value === (e.value = // treat null and undefined the same for the initial value
  i ?? void 0) || // @ts-expect-error
  // `progress` elements always need their value set when it's `0`
  t.value === i && (i !== 0 || t.nodeName !== vt) || (t.value = i ?? "");
}
function Mt(t, i) {
  var e = m(t);
  e.checked !== (e.checked = // treat null and undefined the same for the initial value
  i ?? void 0) && (t.checked = i);
}
function dt(t, i) {
  i ? t.hasAttribute("selected") || t.setAttribute("selected", "") : t.removeAttribute("selected");
}
function L(t, i, e, n) {
  var l = m(t);
  l[i] !== (l[i] = e) && (i === "loading" && (t[j] = e), e == null ? t.removeAttribute(i) : typeof e != "string" && O(t).includes(i) ? t[i] = e : t.setAttribute(i, e));
}
function kt(t, i, e) {
  var n = K, l = X;
  M(null), k(null);
  try {
    // `style` should use `set_attribute` rather than the setter
    i !== "style" && // Don't compute setters for custom elements while they aren't registered yet,
    // because during their upgrade/instantiation they might add more setters.
    // Instead, fall back to a simple "an object, then set as property" heuristic.
    (C.has(t.getAttribute("is") || t.nodeName) || // customElements may not be available in browser extension contexts
    !customElements || customElements.get(t.getAttribute("is") || t.nodeName.toLowerCase()) ? O(t).includes(i) : e && typeof e == "object") ? t[i] = e : L(t, i, e == null ? e : String(e));
  } finally {
    M(n), k(l);
  }
}
function pt(t, i, e, n, l = !1, y = !1) {
  var c = m(t), b = c[G], N = !c[H], f = i || {}, v = t.nodeName === nt;
  for (var A in i)
    A in e || (e[A] = null);
  e.class ? e.class = at(e.class) : e[T] && (e.class = null), e[g] && (e.style ??= null);
  var S = O(t);
  for (const s in e) {
    let o = e[s];
    if (v && s === "value" && o == null) {
      t.value = t.__value = "", f[s] = o;
      continue;
    }
    if (s === "class") {
      var E = t.namespaceURI === "http://www.w3.org/1999/xhtml";
      ct(t, E, o, n, i?.[T], e[T]), f[s] = o, f[T] = e[T];
      continue;
    }
    if (s === "style") {
      ut(t, o, i?.[g], e[g]), f[s] = o, f[g] = e[g];
      continue;
    }
    var _ = f[s];
    if (!(o === _ && !(o === void 0 && t.hasAttribute(s)))) {
      f[s] = o;
      var h = s[0] + s[1];
      if (h !== "_$")
        if (h === "on") {
          const u = {}, p = "_$" + s;
          let a = s.slice(2);
          var d = ot(a);
          if (rt(a) && (a = a.slice(0, -7), u.capture = !0), !d && _) {
            if (o != null) continue;
            t.removeEventListener(a, f[p], u), f[p] = null;
          }
          if (d)
            x(a, t, o), tt([a]);
          else if (o != null) {
            let U = function($) {
              f[s].call(this, $);
            };
            f[p] = et(a, t, U, u);
          }
        } else if (s === "style")
          L(t, s, o);
        else if (s === "autofocus")
          it(
            /** @type {HTMLElement} */
            t,
            !!o
          );
        else if (!b && (s === "__value" || s === "value" && o != null))
          t.value = t.__value = o;
        else if (s === "selected" && v)
          dt(
            /** @type {HTMLOptionElement} */
            t,
            o
          );
        else {
          var r = s;
          N || (r = st(r));
          var w = r === "defaultValue" || r === "defaultChecked";
          if (o == null && !b && !w)
            if (c[s] = null, r === "value" || r === "checked") {
              let u = (
                /** @type {HTMLInputElement} */
                t
              );
              const p = i === void 0;
              if (r === "value") {
                let a = u.defaultValue;
                u.removeAttribute(r), u.defaultValue = a, u.value = u.__value = p ? a : null;
              } else {
                let a = u.defaultChecked;
                u.removeAttribute(r), u.defaultChecked = a, u.checked = p ? a : !1;
              }
            } else
              t.removeAttribute(s);
          else w || S.includes(r) && (b || typeof o != "string") ? (t[r] = o, r in c && (c[r] = D)) : typeof o != "function" && L(t, r, o);
        }
    }
  }
  return f;
}
function Rt(t, i, e = [], n = [], l = [], y, c = !1, b = !1) {
  W(l, e, n, (N) => {
    var f = void 0, v = {}, A = t.nodeName === _t, S = !1;
    if (F(() => {
      var _ = i(...N.map(z)), h = pt(
        t,
        f,
        _,
        y,
        c,
        b
      );
      S && A && "value" in _ && P(
        /** @type {HTMLSelectElement} */
        t,
        _.value
      );
      for (let r of Object.getOwnPropertySymbols(v))
        _[r] || R(v[r]);
      for (let r of Object.getOwnPropertySymbols(_)) {
        var d = _[r];
        r.description === Y && (!f || d !== f[r]) && (v[r] && R(v[r]), v[r] = J(() => ft(t, () => d))), h[r] = d;
      }
      f = h;
    }), A) {
      var E = (
        /** @type {HTMLSelectElement} */
        t
      );
      Q(() => {
        P(
          E,
          /** @type {Record<string | symbol, any>} */
          f.value,
          !0
        ), lt(E);
      });
    }
    S = !0;
  });
}
function m(t) {
  return (
    /** @type {Record<string | symbol, unknown>} **/
    /** @type {any} */
    t[B] ??= {
      [G]: t.nodeName.includes("-"),
      [H]: t.namespaceURI === V
    }
  );
}
var C = /* @__PURE__ */ new Map();
function O(t) {
  var i = t.getAttribute("is") || t.nodeName, e = C.get(i);
  if (e) return e;
  C.set(i, e = []);
  for (var n, l = t, y = Element.prototype; y !== l; ) {
    n = q(l);
    for (var c in n)
      n[c].set && // better safe than sorry, we don't want spread attributes to mess with HTML content
      c !== "innerHTML" && c !== "textContent" && c !== "innerText" && e.push(c);
    l = Z(l);
  }
  return e;
}
export {
  T as CLASS,
  g as STYLE,
  Rt as attribute_effect,
  L as set_attribute,
  Mt as set_checked,
  kt as set_custom_element_data,
  dt as set_selected,
  wt as set_value
};
