// filepond/esm/common/readEntriesFromDataTransfer.js
/*!
* FilePond v5.0.0-beta.41
* Copyright (c) 2017-2026 Pqina B.V.
* Released under the MIT License
* https://filepond.com
*/
import { arrayRemoveFalsy as w } from "../utils/array.js";
import { noop as F } from "../utils/placeholder.js";
import { isFileSystemDirectoryEntry as u, isFunction as f, isFileSystemFileEntry as m } from "../utils/test.js";
import { eachTree as g, mapTreeAsync as A } from "../utils/tree.js";
import { idleCallbackPromise as E } from "../utils/window.js";
async function x(e, r) {
  const { onprogress: i = F, signal: n } = r ?? {}, { items: t } = e ?? {};
  if (!t)
    return [];
  let s = 0, o = 0;
  const h = c(t), l = await y(h, r);
  if (n?.aborted)
    throw n.reason;
  return g(l, (a) => {
    f(a) && s++;
  }), i({ loaded: o, total: s }), await A(
    l,
    async (a) => {
      if (f(a)) {
        if (n?.aborted)
          throw n.reason;
        const p = await a();
        return o++, i({ loaded: o, total: s }), p;
      }
      return a;
    },
    "entries"
  );
}
async function y(e, r) {
  const { signal: i } = r ?? {};
  let n = [];
  for (const t of e) {
    if (!t)
      continue;
    if (i?.aborted)
      throw i.reason;
    let s;
    if (u(t))
      s = {
        name: t.name,
        path: t.fullPath,
        entries: await y(await k(t), r)
      };
    else if (m(t))
      s = async () => {
        const o = await D(t);
        return o.path = t.fullPath, o;
      };
    else
      continue;
    n.push(s);
  }
  return n;
}
function c(e) {
  return Array.from(e).map(b);
}
function T(e) {
  return Array.from(e).map(P);
}
function b(e) {
  return e.webkitGetAsEntry();
}
function P(e) {
  return e.getAsFile();
}
async function k(e) {
  const r = e.createReader(), i = [];
  for (; ; ) {
    const n = await new Promise((t, s) => {
      r.readEntries(t, s);
    });
    if (!n.length)
      break;
    for (const t of n)
      i.push(t);
  }
  return i;
}
async function D(e) {
  return await E(), d(e);
}
async function d(e) {
  return new Promise((r) => e.file(r));
}
function G(e) {
  const r = c(e.items);
  return r.some(u) || r.length > 10;
}
async function L(e) {
  let r = c(e.items);
  if (w(r).length === 0)
    return T(e.items);
  const i = [], n = [];
  for (const t of r)
    m(t) && n.push(d(t));
  return i.push(...await Promise.all(n)), i;
}
export {
  c as dataTransferItemsToEntries,
  T as dataTransferItemsToFiles,
  L as dataTransferToFiles,
  b as getAsEntry,
  P as getAsFile,
  k as readDirectory,
  x as readEntriesFromDataTransfer,
  G as shouldLoadWithIdleCallback
};
