// filepond/esm/svelte/svelte/src/internal/client/dom/operations.js
import { CLASS_CACHE as c, ATTRIBUTES_CACHE as a, STYLE_CACHE as s, TEXT_CACHE as p } from "../constants.js";
import { NAMESPACE_HTML as x } from "../../../constants.js";
import { get_descriptor as r, is_extensible as o } from "../../shared/utils.js";
var _, g, f, u;
function v() {
  if (_ === void 0) {
    _ = window, g = /Firefox/.test(navigator.userAgent);
    var e = Element.prototype, n = Node.prototype, t = Text.prototype;
    f = r(n, "firstChild").get, u = r(n, "nextSibling").get, o(e) && (e[c] = void 0, e[a] = null, e[s] = void 0, e.__e = void 0), o(t) && (t[p] = void 0);
  }
}
function A(e = "") {
  return document.createTextNode(e);
}
// @__NO_SIDE_EFFECTS__
function d(e) {
  return (
    /** @type {TemplateNode | null} */
    f.call(e)
  );
}
// @__NO_SIDE_EFFECTS__
function l(e) {
  return (
    /** @type {TemplateNode | null} */
    u.call(e)
  );
}
function T(e, n) {
  return /* @__PURE__ */ d(e);
}
function h(e, n = !1) {
  {
    var t = /* @__PURE__ */ d(e);
    return t instanceof Comment && t.data === "" ? /* @__PURE__ */ l(t) : t;
  }
}
function S(e, n = 1, t = !1) {
  let i = e;
  for (; n--; )
    i = /** @type {TemplateNode} */
    /* @__PURE__ */ l(i);
  return i;
}
function b(e) {
  e.textContent = "";
}
function y() {
  return !1;
}
function w(e, n, t) {
  return (
    /** @type {T extends keyof HTMLElementTagNameMap ? HTMLElementTagNameMap[T] : Element} */
    document.createElementNS(n ?? x, e, void 0)
  );
}
export {
  _ as _$window,
  T as child,
  b as clear_text_content,
  w as create_element,
  A as create_text,
  h as first_child,
  d as get_first_child,
  l as get_next_sibling,
  v as init_operations,
  g as is_firefox,
  y as should_defer_append,
  S as sibling
};
