// filepond/esm/svelte/svelte/src/internal/client/constants.js
const t = 2, o = 4, s = 8, n = 16777216, c = 16, T = 32, C = 64, A = 128, R = 512, e = 1024, S = 2048, F = 4096, _ = 8192, l = 16384, N = 32768, a = 33554432, r = 65536, D = 131072, O = 262144, I = 524288, m = 1048576, y = 33554432, b = 65536, L = 2097152, Y = 4194304, i = 8388608, d = /* @__PURE__ */ Symbol("$state"), B = /* @__PURE__ */ Symbol("legacy props"), H = /* @__PURE__ */ Symbol(""), g = /* @__PURE__ */ Symbol("attributes"), G = /* @__PURE__ */ Symbol("class"), M = /* @__PURE__ */ Symbol("style"), p = /* @__PURE__ */ Symbol("text"), u = new class extends Error {
  name = "StaleReactionError";
  message = "The reaction that called `getAbortSignal()` was re-run or destroyed";
}(), U = (
  // We gotta write it like this because after downleveling the pure comment may end up in the wrong location
  !!globalThis.document?.contentType && /* @__PURE__ */ globalThis.document.contentType.includes("xml")
);
export {
  Y as ASYNC,
  g as ATTRIBUTES_CACHE,
  c as BLOCK_EFFECT,
  A as BOUNDARY_EFFECT,
  T as BRANCH_EFFECT,
  G as CLASS_CACHE,
  e as CLEAN,
  R as CONNECTED,
  t as DERIVED,
  l as DESTROYED,
  a as DESTROYING,
  S as DIRTY,
  D as EAGER_EFFECT,
  o as EFFECT,
  y as EFFECT_OFFSCREEN,
  I as EFFECT_PRESERVED,
  r as EFFECT_TRANSPARENT,
  i as ERROR_VALUE,
  O as HEAD_EFFECT,
  _ as INERT,
  U as IS_XHTML,
  B as LEGACY_PROPS,
  H as LOADING_ATTR_SYMBOL,
  n as MANAGED_EFFECT,
  F as MAYBE_DIRTY,
  L as REACTION_IS_UPDATING,
  N as REACTION_RAN,
  s as RENDER_EFFECT,
  C as ROOT_EFFECT,
  u as STALE_REACTION,
  d as STATE_SYMBOL,
  M as STYLE_CACHE,
  p as TEXT_CACHE,
  m as USER_EFFECT,
  b as WAS_MARKED
};
