// filepond/esm/svelte/svelte/src/internal/client/render.js
import { init_operations as L, create_text as T } from "./dom/operations.js";
import { array_from as V } from "../shared/utils.js";
import { component_root as j } from "./reactivity/effects.js";
import { TEXT_CACHE as C } from "./constants.js";
import { push as k, component_context as x, pop as z } from "./context.js";
import { boundary as A } from "./dom/blocks/boundary.js";
import { all_registered_events as H, root_event_handles as E, handle_event_propagation as $ } from "./dom/elements/events.js";
import { is_passive_event as N } from "../../utils.js";
let v = !0;
function I(e) {
  v = e;
}
function J(e, r) {
  var n = r == null ? "" : typeof r == "object" ? `${r}` : r;
  n !== /** @type {any} */
  (e[C] ??= e.nodeValue) && (e[C] = n, e.nodeValue = `${n}`);
}
function K(e, r) {
  return P(e, r);
}
const p = /* @__PURE__ */ new Map();
function P(e, { target: r, anchor: n, props: h = {}, events: c, context: g, intro: y = !0, transformError: M }) {
  L();
  var s = void 0, b = j(() => {
    var a = n ?? r.appendChild(T());
    A(
      /** @type {TemplateNode} */
      a,
      {
        pending: () => {
        }
      },
      (i) => {
        k({});
        var o = (
          /** @type {ComponentContext} */
          x
        );
        g && (o.c = g), c && (h._$events = c), v = y, s = e(i, h) || {}, v = !0, z();
      },
      M
    );
    var l = /* @__PURE__ */ new Set(), m = (i) => {
      for (var o = 0; o < i.length; o++) {
        var t = i[o];
        if (!l.has(t)) {
          l.add(t);
          var f = N(t);
          for (const u of [r, document]) {
            var d = p.get(u);
            d === void 0 && (d = /* @__PURE__ */ new Map(), p.set(u, d));
            var w = d.get(t);
            w === void 0 ? (u.addEventListener(t, $, { passive: f }), d.set(t, 1)) : d.set(t, w + 1);
          }
        }
      }
    };
    return m(V(H)), E.add(m), () => {
      for (var i of l)
        for (const f of [r, document]) {
          var o = (
            /** @type {Map<string, number>} */
            p.get(f)
          ), t = (
            /** @type {number} */
            o.get(i)
          );
          --t == 0 ? (f.removeEventListener(i, $), o.delete(i), o.size === 0 && p.delete(f)) : o.set(i, t);
        }
      E.delete(m), a !== n && a.parentNode?.removeChild(a);
    };
  });
  return _.set(s, b), s;
}
let _ = /* @__PURE__ */ new WeakMap();
function O(e, r) {
  const n = _.get(e);
  return n ? (_.delete(e), n(r)) : Promise.resolve();
}
export {
  K as mount,
  I as set_should_intro,
  J as set_text,
  v as should_intro,
  O as unmount
};
